src/Compilers/Core/Portable/Syntax/GreenNode.NodeEnumerable.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis;
+
+internal abstract partial class GreenNode
+{
+    [NonCopyable]
+    public ref struct NodeEnumerable(GreenNode node)
+    {
+        public readonly Enumerator GetEnumerator()
+            => new Enumerator(node);
+
+        [NonCopyable]
+        public ref struct Enumerator
+        {
+            private readonly ArrayBuilder<Syntax.InternalSyntax.ChildSyntaxList.Enumerator> _stack;
+
+            private bool _started;
+            private GreenNode _current;
+
+            public Enumerator(GreenNode node)
+            {
+                _current = node;
+                _stack = ArrayBuilder<Syntax.InternalSyntax.ChildSyntaxList.Enumerator>.GetInstance();
+                _stack.Push(node.ChildNodesAndTokens().GetEnumerator());
+            }
+
+            public readonly void Dispose()
+                => _stack.Free();
+
+            public readonly GreenNode Current
+                => _current;
+
+            public bool MoveNext()
+            {
+                if (!_started)
+                {
+                    // First call that starts the whole process.  We don't actually want to start processing the stack
+                    // yet.  We just want to return the original node (which we already stored into _current).
+                    _started = true;
+                    return true;
+                }
+                else
+                {
+                    while (_stack.TryPop(out var currentEnumerator))
+                    {
+                        while (currentEnumerator.MoveNext())
```

Comment:
Why are we doing a `while` since we never loop back? (we always return `true` when we enter this)

Reply:
true!